### PR TITLE
refactor: use `ScopedArray` (regions) in `MutDeque`

### LINKED ISSUE #3362: Use `ScopedArray` in `MutDeque`

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #3362

- [x] Remove / refactor dependencies (and conversion functions to `MutDeque`)
- [x] Find a solution to `Iterable` instance that expects `Type -> Type` rather than `Type -> Bool -> Type`.

### CODE REVIEW

**Comment by jaschdoc** [comment]
This should work, but the compiler hangs for some reason. I suspect it has something to do with `Array.sameElements` and `MutDeque.sameElements`. I tried constraining the regions to be equal for both parameters but that trick didn't help.

File: main/src/library/MutDeque.flix [hunk 1]
```diff
@@ -27,12 +27,23 @@
...
+
+instance Newable[MutDeque[a]] {
+    pub def new(r: Region[r]): MutDeque[a, r] \ Write(r) = MutDeque.new(r)
+}
+
+instance Scoped[MutDeque[a]] {
+    pub def regionOf(_: MutDeque[a, r]): Region[r] = () as Region[r]
+}
+
+// Find fix for this
```
Comment (magnus-madsen) [inline main/src/library/MutDeque.flix:43]: I think we can start by casting to impure.

Comment (magnus-madsen) [reply main/src/library/MutDeque.flix:43]: Right. We actually need to change the type of Iterable.

I suggest we just wait with the iterable instance, as you have been doing.

Comment (jaschdoc) [reply main/src/library/MutDeque.flix:43]: Roger

**Comment by jaschdoc** [comment]
@magnus-madsen what do we do about that `Iterable` instance? Also, how do we beautify the type of `foldRight`? I feel like `pub def foldRight(f: (a, Unit -> b & ef and r) -> b & ef and r, s: b, d: MutDeque[a, r]): b \ { ef, Read(r) }` could be prettier (especially the type of `f`).

**Comment by magnus-madsen** [comment]
> @magnus-madsen what do we do about that `Iterable` instance? Also, how do we beautify the type of `foldRight`? I feel like

Iterable will need to be updated. I think you can just cast to impure for now or something like this?

> ```
> pub def foldRight(f: (a, Unit -> b & ef and r) -> b & ef and r, s: b, d: MutDeque[a, r]): b \ { ef, Read(r) }
> ```

I am not sure, but I don't think `f` should have effect `r`, only `ef`.

**Comment by jaschdoc** [comment]
That throws a type error, though. I'll post it in a sec.

**Comment by jaschdoc** [comment]
```flix
pub def foldRight(f: (a, Unit -> b & ef) -> b & ef, s: b, d: MutDeque[a, r]): b \ { ef, Read(r) } =
    let MutDeque(ar_, front_, back_) = d;
    let ar = deref ar_;
    let front = deref front_;
    let back = deref back_;
    let c = capacity(d) - 1;
    def loop(i, e) =
        if (i == e)
            s
        else {
            f(ar[i], _ -> loop((i + 1) &&& c, e))
        };
    loop(front, back)
```
errors with:
```
Unable to compile. Failed with: 1 errors.

-- Type Error -------------------------------------------------- MutDeque.flix

>> The type scheme: '∀(a, region, b, eff). (a -> ((Unit ->{eff and region} b) ->{eff and region} b)) -> (b -> (MutDeque[a, r] ->{region and eff} result))' cannot be generalized to '∀(a, ef, b, r). (a -> ((Unit ->{ef} b) ->{ef} b)) -> (b -> (MutDeque[a, r] ->{ef and r} b))'.

183 |     pub def foldRight(f: (a, Unit -> b & ef) -> b & ef, s: b, d: MutDeque[a, r]): b \ { ef, Read(r) } =
                  ^^^^^^^^^
                  unable to generalize the type scheme.

The declared type does not match the inferred type:

  Declared: ∀(a, ef, b, r). (a -> ((Unit ->{ef} b) ->{ef} b)) -> (b -> (MutDeque[a, r] ->{ef and r} b))
  Inferred: ∀(a, region, b, eff). (a -> ((Unit ->{eff and region} b) ->{eff and region} b)) -> (b -> (MutDeque[a, r] ->{region and eff} result))
```

**Comment by magnus-madsen** [comment]
OK, I think agree with the signature of foldRight. I will think about how  to make it more pretty, but I think it is correct for now.

**Comment by jaschdoc** [comment]
> OK, I think agree with the signature of foldRight. I will think about how to make it more pretty, but I think it is correct for now.

Alright, sounds good.

**Comment by jaschdoc** [comment]
> Iterable will need to be updated. I think you can just cast to impure for now or something like this?

I'm not sure we can simply cast to pure since the `Iterable` type class expects a different kind. Specifically:
```flix
instance Iterable[MutDeque] {
    pub def iterator(t: MutDeque[a, Impure]): Iterator[a] & Impure =
        let MutDeque(a, f, b) = t;
        let len = Array.length(deref a);
        let copy = deref a;
        let i = f;
        let done = () -> (deref i) == deref b;
        let next = () -> {
            if (deref i != deref b) {
                let j = {
                    if (deref i == len) 0
                    else deref i
                };
                let x = copy[j];
                i := j + 1;
                x
            } else {
                bug!("Empty Iterator!")
            }
        };
        Iterator(done, next)
}
```
errors with:
```

Unable to compile. Failed with: 2 errors.

-- Kind Error -------------------------------------------------- MutDeque.flix

>> Expected kind 'Type -> Type' here, but kind 'Type -> Bool -> Type' is used.

43 | instance Iterable[MutDeque] {
                       ^^^^^^^^
                       unexpected kind.

Expected kind: Type -> Type
Actual kind:   Type -> Bool -> Type
```

**Comment by jaschdoc** [comment]
We could add an instance for an opaque type such as `opaque type StaticMutDeque[a] = MutDeque[a, Impure]`

**Comment by magnus-madsen** [comment]
Lets delete the Iterable instances for now. We have to refactor Iterable anyway.

**Comment by jaschdoc** [comment]
Roger

**Comment by magnus-madsen** [comment]
I suggest to do that in a separate PR.

**Comment by magnus-madsen** [comment]
Very nice. So now we have one more data structures with regions 👍

**Pull request merged**